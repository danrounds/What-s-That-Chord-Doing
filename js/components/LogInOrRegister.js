import React from 'react';
import {connect} from 'react-redux';
import {Router, Redirect, hashHistory} from 'react-router';

import NavBar from './NavBar';
import * as actions from '../actions';

export class LogInOrRegister extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            register: false,
            statusText: '',
        };
        this.onLogIn = this.onLogIn.bind(this);
        this.onRegister = this.onRegister.bind(this);
    }

    onLogIn() {
        this.setState({ register: false });
        this.props.dispatch(actions.getUserScores(this.name.value, this.password.value));
        return false;           // Keeps page from refreshing
    }

    onRegister() {
        if (!this.state.register) {
            this.setState({ register: true, statusText: '' });
        } else {
            try {
                const name = this.name.value, pw = this.password.value,
                      pw_ = this.passwordConfirm.value;

                if (name && pw === pw_) {
                    const matchingRegEx = /[a-zA-Z0-9_]+/.exec(name) || [];
                    if (matchingRegEx[0] !== name)
                        this.setState({ statusText: 'Username should be letters, numbers, and underscores' });
                    else if (pw.length < 6)
                        this.setState({ statusText: 'Please choose a password of at least six characters'});
                    else
                        this.props.dispatch(actions.makeUserAccount(name, pw));
                } else if (pw !== pw_)
                    this.setState({ statusText: 'Passwords don\'t match'});
            } catch(e) {;;;;}
        }
        return false;           // Keeps page from refreshing
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.api.error !== this.props.api.error) {
            if (nextProps.api.error === 401)
                this.setState({ statusText: 'Invalid password or username' });
            else if (nextProps.api.error === 409)
                this.setState({ statusText: 'Username already exists; try a new one' });
            else if (nextProps.api.error)            
                this.setState({ statusText: 'Submit failed; try again' });
        }
    }

    render() {
        if (this.props.api.userScores.name) {
            return (
                <Router history={hashHistory}>
                  <Redirect from="log-in-or-register" to="/" />
                </Router>
            );
        }

        return (
            <div>
              <NavBar parent="LogIn"/>

              <div className="log-in-status">{this.state.statusText}</div>

              <form className="log-in-and-registration-form">
                <label>
                  Name<br/>
                  <input ref={el => { this.name = el; }}
                    type="text" autoFocus required /><br/>
                </label>
                <label>
                  Password<br/>
                  <input ref={el => { this.password = el; }}
                    type="password" required /><br/>
                </label>

                {this.state.register ?
                    (<label>Confirm password<br/>
                      <input ref={el => { this.passwordConfirm = el; }}
                            type="password" required /><br/>
                     </label>)
                 : null}

                <button onClick={this.onLogIn}>Log in</button>
                <button onClick={this.onRegister}>Register</button>
              </form>

            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    api: state.api,
});

export default connect(mapStateToProps)(LogInOrRegister);
