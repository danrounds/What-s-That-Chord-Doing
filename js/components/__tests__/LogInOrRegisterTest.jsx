import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { LogInOrRegister } from '../LogInOrRegister';

// Testing the various permutations of state that our LogIn... component might
// encounter. Really just testing rendering, here

describe('Score-based components', () => {
    // Success:
    test('LogInOrRegister, when log-in succeeds', () => {
        const comp = shallow(<LogInOrRegister api={{ authToken: 'some_string' }}/>);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('LogInOrRegister, when registration succeeds', () => {
        const comp = shallow(<LogInOrRegister api={{ authToken: 'some_string' }}/>);
        comp.setState({ register: true });
        expect(toJson(comp)).toMatchSnapshot();
    });    

    // Pending:
    test('LogInOrRegister, when logging in & pending', () => {
        const comp = shallow(<LogInOrRegister api={{}}/>);
        comp.setProps({ api: { pending: true }});
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('LogInOrRegister, when registering & pending', () => {
        const comp = shallow(<LogInOrRegister api={{}} />);
        comp.setState({ register: true });
        comp.setProps({ api: { pending: true }});
        expect(toJson(comp)).toMatchSnapshot();
    });

    // Errors:
    test('LogInOrRegister, when logging in & username doesn\'t exist', () => {
        const comp = shallow(<LogInOrRegister api={{ pending: true }}/>);
        comp.setProps({ api: { error: 404 }});
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('LogInOrRegister, when logging in & password is wrong', () => {
        const comp = shallow(<LogInOrRegister api={{ pending: true }}/>);
        comp.setProps({ api: { error: 401 }});
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('LogInOrRegister, when registering & username already exists', () => {
        const comp = shallow(<LogInOrRegister api={{ pending: true }}/>);
        comp.setState({ register: true });
        comp.setProps({ api: { error: 409 }});
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('LogInOrRegister, when server doesn\'t respond', () => {
        const comp = shallow(<LogInOrRegister api={{ pending: true }}/>);
        comp.setProps({ api: { error: 'other' }});
        expect(toJson(comp)).toMatchSnapshot();
    });
});
