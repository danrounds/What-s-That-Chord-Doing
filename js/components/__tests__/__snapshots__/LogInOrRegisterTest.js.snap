// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score-based components LogInOrRegister, when log-in succeeds 1`] = `
<Router
  history={
    Object {
      "createHref": [Function],
      "createKey": [Function],
      "createLocation": [Function],
      "createPath": [Function],
      "getCurrentLocation": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "listenBefore": [Function],
      "push": [Function],
      "replace": [Function],
      "transitionTo": [Function],
    }
  }
  render={[Function]}
>
  <Redirect
    from="log-in-or-register"
    to="/"
  />
</Router>
`;

exports[`Score-based components LogInOrRegister, when logging in & password is wrong 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status log-in-err"
  >
    Invalid password
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;

exports[`Score-based components LogInOrRegister, when logging in & pending 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status"
  >
    Checking...
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;

exports[`Score-based components LogInOrRegister, when logging in & username doesn't exist 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status log-in-err"
  >
    Username doesn't exist
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;

exports[`Score-based components LogInOrRegister, when registering & pending 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status"
  >
    Registering...
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <label>
      Confirm password
      <br />
      <input
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;

exports[`Score-based components LogInOrRegister, when registering & username already exists 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status log-in-err"
  >
    Username already exists; try a new one
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <label>
      Confirm password
      <br />
      <input
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;

exports[`Score-based components LogInOrRegister, when registration succeeds 1`] = `
<Router
  history={
    Object {
      "createHref": [Function],
      "createKey": [Function],
      "createLocation": [Function],
      "createPath": [Function],
      "getCurrentLocation": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "listenBefore": [Function],
      "push": [Function],
      "replace": [Function],
      "transitionTo": [Function],
    }
  }
  render={[Function]}
>
  <Redirect
    from="log-in-or-register"
    to="/"
  />
</Router>
`;

exports[`Score-based components LogInOrRegister, when server doesn't respond 1`] = `
<div
  onKeyDown={[Function]}
>
  <Connect(NavBar)
    parent="LogIn"
  />
  <div
    className="log-in-status log-in-err"
  >
    Submit failed; try again
  </div>
  <form
    action="javascript:void(0)"
    className="log-in-and-registration-form"
  >
    <label>
      Name
      <br />
      <input
        autoFocus={true}
        required={true}
        type="text"
      />
      <br />
    </label>
    <label>
      Password
      <br />
      <input
        className="log-in-input"
        required={true}
        type="password"
      />
      <br />
    </label>
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid green",
          "color": "#333",
        }
      }
    >
      Log in
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "gray",
        }
      }
    >
      Register
    </button>
  </form>
</div>
`;
